renderdoc/driver/gl/gl_pixelhistory.cpp

```diff
@@ -0,0 +1,225 @@
+          usage == ResourceUsage::GenMips);
+}
+
+struct GLPixelHistoryResources
+{
+  // Used for offscreen rendering for draw call events.
+  GLuint colorImage;
+  GLuint dsImage;
+  GLuint frameBuffer;
+  GLuint occlusionQuery;
+};
+
+bool PixelHistorySetupResources(WrappedOpenGL *driver, GLPixelHistoryResources &resources,
+                                const TextureDescription &desc, const Subresource &sub,
+                                uint32_t numEvents)
+{
+  // Allocate a framebuffer that will render to the textures
+  driver->glGenFramebuffers(1, &resources.frameBuffer);
+  driver->glBindFramebuffer(eGL_FRAMEBUFFER, resources.frameBuffer);
+
+  // Allocate a texture for the pixel history colour values
+  driver->glGenTextures(1, &resources.colorImage);
+  driver->glBindTexture(eGL_TEXTURE_2D, resources.colorImage);
+  driver->CreateTextureImage(resources.colorImage, eGL_RGBA32F, eGL_NONE, eGL_NONE, eGL_TEXTURE_2D,
+                             2, desc.width >> sub.mip, desc.height >> sub.mip, 1, desc.msSamp, 1);
+  driver->glFramebufferTexture(eGL_FRAMEBUFFER, eGL_COLOR_ATTACHMENT0, resources.colorImage, 0);
+
+  // Allocate a texture for the pixel history depth/stencil values
+  driver->glGenTextures(1, &resources.dsImage);
+  driver->glBindTexture(eGL_TEXTURE_2D, resources.dsImage);
+  driver->CreateTextureImage(resources.dsImage, eGL_DEPTH32F_STENCIL8, eGL_NONE, eGL_NONE,
+                             eGL_TEXTURE_2D, 2, desc.width >> sub.mip, desc.height >> sub.mip, 1,
+                             desc.msSamp, 1);
+  driver->glFramebufferTexture(eGL_FRAMEBUFFER, eGL_DEPTH_STENCIL_ATTACHMENT, resources.dsImage, 0);
+
+  driver->glGenQueries(1, &resources.occlusionQuery);
+
+  return true;
+}
+
+bool PixelHistoryDestroyResources(WrappedOpenGL *driver, const GLPixelHistoryResources &resources)
+{
+  driver->glDeleteTextures(1, &resources.colorImage);
+  driver->glDeleteTextures(1, &resources.dsImage);
+  driver->glDeleteFramebuffers(1, &resources.frameBuffer);
+
+  driver->glDeleteQueries(1, &resources.occlusionQuery);
+
+  return true;
+}
+
+rdcarray<EventUsage> QueryModifyingEvents(WrappedOpenGL *driver, GLPixelHistoryResources &resources,
+                                          const rdcarray<EventUsage> &events, int x, int y)
+{
+  driver->glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT | GL_STENCIL_BUFFER_BIT);
+  driver->ReplayLog(0, events[0].eventId, eReplay_WithoutDraw);
+
+  rdcarray<EventUsage> modEvents;
+
+  for(size_t i = 0; i < events.size(); i++)
+  {
+    if(events[i].usage == ResourceUsage::Clear || isDirectWrite(events[i].usage))
+    {
+      modEvents.push_back(events[i]);
+    }
+    else
+    {
+      driver->glBindFramebuffer(eGL_FRAMEBUFFER, resources.frameBuffer);
+
+      driver->glDisable(eGL_DEPTH_TEST);
+      driver->glDisable(eGL_STENCIL_TEST);
+      driver->glEnable(eGL_SCISSOR_TEST);
+      driver->glScissor(x, y, 1, 1);
```

Comment:
There are a couple more state bits that need to be disabled here (e.g. backface culling, sample masking) which could affect pixels being written to.